File: frontend/lib/services/google_maps_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../core/constants/api_constants.dart';

/// Service to call Google Maps API through backend (secure)
class GoogleMapsService {
  /// Get authentication token
  static Future<String?> _getAuthToken() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      return prefs.getString('auth_token');
    } catch (e) {
      return null;
    }
  }

  /// Reverse geocoding - Get address from coordinates
  /// Calls: GET /api/google-maps/reverse-geocode?lat={lat}&lng={lng}
  static Future<Map<String, dynamic>?> getAddressFromCoords(
    double lat,
    double lng,
  ) async {
    try {
      final token = await _getAuthToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final url = Uri.parse(
        '${ApiConstants.baseUrl}/google-maps/reverse-geocode?lat=$lat&lng=$lng',
      );

      final response = await http.get(
        url,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success'] == true) {
          return data['data'] as Map<String, dynamic>;
        }
      }

      return null;
    } catch (e) {
      print('Error getting address from coordinates: $e');
      return null;
    }
  }

  /// Places Autocomplete - Search for places
  /// Calls: GET /api/google-maps/places/autocomplete?query={query}&country={country}
  static Future<Map<String, dynamic>?> searchPlaces(
    String query, {
    String country = 'pk',
  }) async {
    try {
      final token = await _getAuthToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final encodedQuery = Uri.encodeComponent(query);
      final url = Uri.parse(
        '${ApiConstants.baseUrl}/google-maps/places/autocomplete?query=$encodedQuery&country=$country',
      );

      final response = await http.get(
        url,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success'] == true) {
          return data['data'] as Map<String, dynamic>;
        }
      }

      return null;
    } catch (e) {
      print('Error searching places: $e');
      return null;
    }
  }

  /// Place Details - Get details for a specific place
  /// Calls: GET /api/google-maps/places/details?placeId={placeId}&fields={fields}
  static Future<Map<String, dynamic>?> getPlaceDetails(
    String placeId, {
    String fields = 'geometry',
  }) async {
    try {
      final token = await _getAuthToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final encodedPlaceId = Uri.encodeComponent(placeId);
      final url = Uri.parse(
        '${ApiConstants.baseUrl}/google-maps/places/details?placeId=$encodedPlaceId&fields=$fields',
      );

      final response = await http.get(
        url,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success'] == true) {
          return data['data'] as Map<String, dynamic>;
        }
      }

      return null;
    } catch (e) {
      print('Error getting place details: $e');
      return null;
    }
  }
}

File: backend/utils/googleMapsService.js
// Google Maps API Service - Backend proxy for secure API key usage
import axios from "axios";

/**
 * Reverse geocoding - Get address from coordinates
 * @param {number} lat - Latitude
 * @param {number} lng - Longitude
 * @returns {Promise<Object|null>} Google Geocoding API response
 */
export const getAddressFromCoords = async (lat, lng) => {
  try {
    const apiKey = process.env.GOOGLE_MAPS_API_KEY;
    
    if (!apiKey) {
      console.error("❌ GOOGLE_MAPS_API_KEY not set in .env");
      return null;
    }

    const url = `https://maps.googleapis.com/maps/api/geocode/json?latlng=${lat},${lng}&key=${apiKey}&language=en`;
    
    const response = await axios.get(url);
    
    return response.data;
  } catch (error) {
    console.error("Google Geocoding API error:", error.message);
    return null;
  }
};

/**
 * Places Autocomplete - Search for places
 * @param {string} query - Search query
 * @param {string} country - Country code (default: 'pk' for Pakistan)
 * @returns {Promise<Object|null>} Google Places Autocomplete response
 */
export const searchPlaces = async (query, country = "pk") => {
  try {
    const apiKey = process.env.GOOGLE_MAPS_API_KEY;
    
    if (!apiKey) {
      console.error("❌ GOOGLE_MAPS_API_KEY not set in .env");
      return null;
    }

    const encodedQuery = encodeURIComponent(query);
    const url = `https://maps.googleapis.com/maps/api/place/autocomplete/json?input=${encodedQuery}&key=${apiKey}&language=en&components=country:${country}`;
    
    const response = await axios.get(url);
    
    return response.data;
  } catch (error) {
    console.error("Google Places Autocomplete API error:", error.message);
    return null;
  }
};

/**
 * Place Details - Get details for a specific place
 * @param {string} placeId - Google Place ID
 * @param {string} fields - Comma-separated fields to return (default: 'geometry')
 * @returns {Promise<Object|null>} Google Place Details response
 */
export const getPlaceDetails = async (placeId, fields = "geometry") => {
  try {
    const apiKey = process.env.GOOGLE_MAPS_API_KEY;
    
    if (!apiKey) {
      console.error("❌ GOOGLE_MAPS_API_KEY not set in .env");
      return null;
    }

    const url = `https://maps.googleapis.com/maps/api/place/details/json?place_id=${placeId}&key=${apiKey}&fields=${fields}`;
    
    const response = await axios.get(url);
    
    return response.data;
  } catch (error) {
    console.error("Google Place Details API error:", error.message);
    return null;
  }
};

